### PR TITLE
KT-58547 - [IC] Relocatable IC caches for projects with custom buildDir

### PR DESCRIPTION
**Pull request by hungvietnguyen**

### CODE REVIEW

File: compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/storage/SourceToOutputFilesMapTest.kt [hunk 1]
```diff
@@ -7,129 +7,92 @@ package org.jetbrains.kotlin.incremental.storage
...
-    }
 
-    @After
-    override fun tearDown() {
-        stofMap.flush(false)
-        stofMap.closeForTest()
-        super.tearDown()
+        fooDotKt = srcDir.resolve("Foo.kt")
+        fooDotClass = classesDir.resolve("Foo.class")
     }
```
Comment (gavra0) [inline compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/storage/SourceToOutputFilesMapTest.kt:42]: Should we still close the map to make sure we don't get issues on Windows?

File: build-common/test/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverterTest.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.incremental.storage
+
+import org.jetbrains.kotlin.TestWithWorkingDir
+import org.junit.Test
+import kotlin.test.assertFailsWith
+
+class RelocatableFileToPathConverterTest : TestWithWorkingDir() {
```
Comment (gavra0) [inline build-common/test/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverterTest.kt:12]: Use Junit rule, `TemporaryFolder`?

File: build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * (https://docs.gradle.org/current/userguide/build_cache_concepts.html#relocatability).
+ */
+class RelocatableFileToPathConverter(private val baseDir: File) : FileToPathConverter {
+
+    private val unixStyleBaseDirPathPrefix = "${baseDir.invariantSeparatorsPath}/"
+
+    override fun toPath(file: File): String {
+        check(file.invariantSeparatorsPath.startsWith(unixStyleBaseDirPathPrefix)) {
+            "The given file '${file.path}' is located outside the base directory '${baseDir.path}'"
+        }
```
Comment (gavra0) [inline build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt:24]: Can compiler plugins  output .class files directly? If so, can they specify a custom location and will those classes still be processed by the IC caches?

File: build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        return file.relativeTo(baseDir).invariantSeparatorsPath
+    }
+
+    override fun toFile(path: String): File {
+        // Note: The given path is Unix-style but baseDir could be Windows-style; call normalize() so that the style of the returned file's
+        // path is consistent with baseDir.
+        return baseDir.resolve(path).normalize()
```
Comment (gavra0) [inline build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt:31]: To ensure consistency, should we throw if path is absolute?

File: compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/storage/SourceToOutputFilesMapTest.kt (hunk 1 above)
Comment (hungvietnguyen) [reply compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/storage/SourceToOutputFilesMapTest.kt:42]: Ah yes. I've also made `LazyStorage` implement `Closeable` now to emphasize this.

File: build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt (hunk 3 above)
Comment (hungvietnguyen) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt:24]: If the compiler plugins generate `.class` files outside `classesDir`, the IC will not be aware of those (there's currently no API for the two to communicate on this info), so those classes will not be processed by the IC caches.

File: build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt (hunk 4 above)
Comment (hungvietnguyen) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt:31]: Yes if `toFile` is an independent API.

However, here `toFile()` is supposed to be called only on the returned value of a `toPath()`, so given `toPath()` is implemented correctly, we will always get a Unix-style relative path here.

If we add this check, we might need to also check that the path is Unix-style, and add similar checks for `baseDir` to and the `file` argument in `toPath()`, so we'll end up having a lot of checks.

File: build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt (hunk 3 above)
Comment (gavra0) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt:24]: Ok, then this sounds like a safe assertion.

File: build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt (hunk 4 above)
Comment (gavra0) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt:31]: Kk, leaving it as-is sounds good.

**Review by gavra0 — COMMENTED** [review]
LGTM

**Comment by ALikhachev** [comment]
LGTM, will be merged via our Space instance

**Review by ALikhachev — APPROVED** [review]

**Comment by ALikhachev** [comment]
Merged into `master` as https://github.com/JetBrains/kotlin/commit/894ba9ab809c400de048d43fa98f89087100fcbc

**Pull request closed without merging**